### ISSUE TITLE
Router issues on distribution

### ISSUE THREAD

**Issue by stephenhandley** [redwoodjs/graphql#4689]
Running into a bug when updating from `0.39.4` to `0.41.0` where when changing routes via `NavLink`, the path (and nav selected path update) but the associated page for that route doesn't update / render and the previous page is still shown. This doesn't happen when developing locally, only on deployed app. I can duplicate locally when running `Servez` on the dist folder output by `yarn rw deploy render web`.

Can duplicate the issue every time after page reload. Things work whenever navigating to a page from the root homepage but navigating between pages doesn't work i.e. 
| from | to | |
| --- | --- | --- |
|/ | /foo | ✅  |
|/foo | / | ✅  |
|/ | /bar | ✅  |
|/bar| / | ✅  |
|/bar | /foo | ❌  |
|/foo | /bar | ❌  |

**Comment by simoncrypta** [comment]
Hi @stephenhandley, 

Thanks for reporting this issue! :)
I cannot replicate the issue in with the last version of Redwood on Render. Can you confirm if it is also working for you at V0.49.1 ?

**Comment by thedavidprice** [comment]
Hmm, here's my deployed Redwood Test-project running on Render — it's updated with every canary release. Currently cannot replicate this issue either. Seems to be working correct here, yes?
https://render-postgres-web.onrender.com

**Comment by stephenhandley** [comment]
@simoncrypta @thedavidprice because of this issue, we aren't at a point of being able to upgrade due to https://github.com/redwoodjs/redwood/issues/4597#issuecomment-1058490585 but I will follow up again on reproduction once the magic fix lands or we switch to clerk..

**Comment by thedavidprice** [comment]
@stephenhandley Understood and that's super frustrating — I'm with you. Just wish I could do more than direct traffic. But we're heads down on getting out the final RC for v1. Lots more to come. Just hang in there!

@shahbaz17 please let us know if there's any way we can help support you here. Shared Slack channel would be welcome if that could help with communication. Thank you in advance!

**Comment by stephenhandley** [comment]
@thedavidprice just upgraded to `v1.0.0-rc.final.3` and still seeing this issue. 

Worth noting that the pages in here are explicitly imported in the router and are in subdirectories so perhaps related to 
https://github.com/redwoodjs/redwood/issues/4895
https://github.com/redwoodjs/redwood/issues/4896

**Comment by stephenhandley** [comment]
Still seeing this in 1.0.0

**Comment by stephenhandley** [comment]
@jtoar thanks for following up, giving this a go at reproing with your repo, but haven't had any luck yet. there's a few nonstandard things going on in our codebase in terms of how we determine routes, particularly in that we are running both an app and brand site off the same codebase with separate routes for each. Not seeing any issues with our App, just our brand site, which is the one place where we aren't using auth. 

Basically we see the url update but the page doesn't rerender. 

One new curious thing I didn't see in 0.41 but do see in 1.0.0 in our repo is that the routing is now also broken when running `yarn rw dev`, it used to just be on dist, but now seeing it both places

**Comment by stephenhandley** [comment]
we just changed from doing a default unnamed export to named exports and that fixes things when running things via `yarn rw dev`

i.e. 
// AnotherPage.tsx
```
export default function () {
```
to 
```
export default function AnotherPage() {
```

**Comment by stephenhandley** [comment]
@jtoar https://github.com/jtoar/router-repro/pull/1

**Comment by stephenhandley** [comment]
Also an example where `yarn rw dev` works but not `yarn rw deploy render web` https://github.com/jtoar/router-repro/pull/2/files

**Comment by stephenhandley** [comment]
This is not ideal, but flattening out the Routes into a single file fixes the dist issue
https://github.com/stephenhandley/router-repro/pull/1

To be honest, I really wish there weren't any magic imports in Redwood, i'd rather have to manage the imports by hand and not get bitten by things like this.

**Comment by thedavidprice** [comment]
> To be honest, I really wish there weren't any magic imports in Redwood, i'd rather have to manage the imports by hand and not get bitten by things like this.

@dac09 isn't there a way to "escape" the magic here when projects require overriding?

**Comment by dac09** [comment]
Hey - sorry you were struggling with this - and I hadn't spotted this before!

We do not advise splitting your Routes into multiple files in Redwood Router. This was a concious decision on our side - but we definitely need to make this clearer in the documentation.

If you come from the React Router or React Native world, it might feel like the natural thing to do - but often going doing this path leads to "untracable" routes. It seems counter intuitive at first, but having one file describe all the available routes means fewer hoops to jump when you're trying to debug why something is routing oddly. 

> @dac09 isn't there a way to "escape" the magic here when projects require overriding?
Queing up for discussing at Core team meeting DP! We had this discussion maybe a year ago, but I think we should revisit this and see how we can improve the DX here.

**Comment by stephenhandley** [comment]
@dac09 is the reason for not allowing multiple route files because of magic imports or is there some other factor at play there?

**Comment by jtoar** [comment]
Hey @stephenhandley I brought this one up to the core team today during one of our meetings. We want to know more about your use case: are you splitting up your routes this way purely for organizational reasons? Or is there another benefit to it, like improved performance in some way?

Also, half the reason Redwood automatically imports pages for you is so that it can code split automatically. If you manually import pages into the Routes file, they're included in the bundle. So if you end up doing it all yourself you'd have to make sure to lazily import every page. Just a lot more overhead!

**Comment by stephenhandley** [comment]
@jtoar I hear you on the bundle size bloat without lazy imports. That makes sense.

In the example recreation I linked above, it shows how we have different web side "apps" we run based on the subdomain. The split was primarily organizational: we put pages into separate folders per app and an associated routes file for each. 

There were a couple gotchas I ran into when trying to organize things that way our codebase, for example the framework expecting the `Routes.tsx` file to exist at a predetermined path and similarly for `db.ts` on the api side. Things blew up during the build with opaque errors and had to trial and error through that. Not a huge deal, but was a bit confusing at first.

I suppose some of that could be solved via documentation i.e. some section on code folder structure / organization etc. that lists out files users shouldn't move, and what is fair game.

That said, this is not a hack we're particularly happy about, but it was a workaround needed since Redwood doesn't yet support multiple frontend sides. If we had something like this where each `web-foo` and `web-bar` had their own dedicated apps, pages, routes and then we had some `web-shared` component package to share components / helpers / hooks etc, that would probably be a cleaner solution overall
```
/web-foo
/web-bar
/web-shared
/api 
```

For what its worth, my preference for minimizing framework magic stems back from having being bitten by it on so many other frameworks over the years. I think what often is presented as a convenience can end up being a source of confusion and cognitive load in devex since people have to keep these rules / exceptions in mind. Keeping it simple and requiring me to write in the imports by hand requires more work up front, but keeps the model simpler which I'd trade for without hesitation. Sure, maybe that means I have to edit the Routes.tsx file by hand after generating a page, but the reality is that even with the current setup, I usually end up having to do so in order to place the added route within the proper Set/Layout since it just gets put in at the top level.

**Comment by thedavidprice** [comment]
> That said, this is not a hack we're particularly happy about, but it was a workaround needed since Redwood doesn't yet support multiple frontend sides.

This is really interesting. A lot of what people mean when they request "multiple frontend sides" is integration support for Next or Svelte or etc. But you're saying you want to be able to use Redwood Web for multiple apps at the same time, which really just means you want to be able to do something like this via build:
```
web/dist/app1
web/dist/app2
```

Right?

Spitballing here — what if we mapped appX <> `Routes.appX.tsx`<> `App.AppX.tsx` and then ran a build against each respective App and Route file set? You'd be able to keep all the code within web/ and share as needed. There'd be a lot of local dev and overall config considerations to evaluate, but this feels near-term accessible.

**Comment by chenkie** [comment]
@thedavidprice I have a use case for what you described above. +1 for multiple Redwood web sides 🙂

**Comment by Tobbe** [comment]
I have a main RW /web side right now for customer facing sides.
I'd like to add a second RW web side for an admin panel (ideally under /admin next to /web) served at mysite.com/admin or admin.mysite.com
My main site uses i18n, my admin site wouldn't need that. I'd like to use different ui toolkits for /web and /admin (/web is fully custom css, /admin would be something standard like stock Chakra maybe). I'd want the sides to be as separated as possible so that I'm guaranteed that if I accidentally break something for the admin side it can never affect the web side, and vice versa.

**Comment by Tobbe** [comment]
@stephenhandley and everyone else: There are some exciting experiments happening over at the RW forums that you might be interested in

https://community.redwoodjs.com/t/multiple-redwoodjs-in-monorepo/3239

Please chime in :)
